### ISSUE TITLE
SaveManager does not close project when reading metadata fails while restoring workspace

### ISSUE THREAD

**Issue by HeikoKlare** [eclipse-platform/eclipse.platform#377]
When a workspace is not properly saved while/after performing resource changes, inconsistent metadata can lead to a faulty state when restoring the workspace during subsequent loads.

In detail: The test `TestBug294854.testDelete` randomly fails (see https://ci.eclipse.org/platform/job/eclipse.platform/job/PR-371/6/testReport/ and particularly the `FileNotFoundException` in the according log https://ci.eclipse.org/platform/job/eclipse.platform/job/PR-371/6/consoleFull).
The test deletes a projects and afterwards calls `System.exit`. After a project deletion, the snapshot job is scheduled to store the workspace modification. However, there is a race condition because the snapshot job may properly write data before `System.exit` is executed or after it is executed. For that reason the other tests in `TestBug294854` try to wait for the snapshot job being finished, as has been discussed here: https://bugs.eclipse.org/bugs/show_bug.cgi?id=294854
When the snapshot job does not finish before calling `System.exit`, the subsequent workspace restoration by the `SaveManager` tries to load the deleted project and throws a `FileNotFoundException`. Although the exception is caught, the project is not closed as intended when an exception is thrown due to a missing try-finally:
https://github.com/eclipse-platform/eclipse.platform/blob/8f9546dbcdbef9ad236c79821c030cad39ec250a/resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/SaveManager.java#L914-L916

In summary, I suggest the following changes to fix the issues:
* `SaveManager` should properly close the project when an exception is thrown while writing metadata on workspace restoration
* `TestBug294854.testDelete` has to be fixed and should be split into two test cases
  1. Wait for the snapshot job to be finished and expect the project to be properly removed afterwards
  1. Call `System.exit` deterministically between deleting the project and executing the snapshot job and expect the project to be closed when restoring the workspace
